crates/unftp-auth-jsonfile/src/lib.rs

```diff
@@ -186,18 +215,35 @@ impl JsonFileAuthenticator {
             } => verify(PBKDF2_HMAC_SHA256, *pbkdf2_iter, pbkdf2_salt, given_password.as_bytes(), pbkdf2_key).map_err(|_| ()),
         }
     }
+
+    fn ip_ok(creds: &libunftp::auth::Credentials, actual_creds: &UserCreds) -> bool {
+        match &actual_creds.ip_ranges {
+            Some(allowed) => match creds.source_ip {
+                std::net::IpAddr::V4(ref ip) => allowed.contains(ip),
+                _ => false,
+            },
+            None => true,
+        }
+    }
 }
 
 #[async_trait]
 impl Authenticator<DefaultUser> for JsonFileAuthenticator {
     #[tracing_attributes::instrument]
-    async fn authenticate(&self, username: &str, given_password: &str) -> Result<DefaultUser, AuthenticationError> {
-        if let Some(actual_password) = self.credentials_map.get(username) {
-            if let Ok(()) = Self::check_password(given_password, actual_password) {
-                return Ok(DefaultUser);
-            } else {
-                sleep(Duration::from_millis(1500)).await;
-                return Err(AuthenticationError::BadPassword);
+    async fn authenticate(&self, username: &str, creds: &libunftp::auth::Credentials) -> Result<DefaultUser, AuthenticationError> {
+        if let Some(actual_creds) = self.credentials_map.get(username) {
+            match creds.password {
+                None => Err(AuthenticationError::BadPassword),
+                Some(ref given_password) => {
+                    if Self::check_password(given_password, &actual_creds.password).is_ok() {
+                        Self::ip_ok(creds, &actual_creds)
+                            .then(|| Ok(DefaultUser {}))
+                            .unwrap_or(Err(AuthenticationError::IpDisallowed))
```

Comment:
Securitywise for a IP mismatch it's preferable to let the actor receive a "bad password treatment": letting him wait for 1,5 seconds. The client should be none the wiser as to whether the password was indeed bad or whether the IP was not allowed.

Reply:
I'll add the 1,5 seconds. For the error type. Did we not previously decide that policy decisions sit with the library user i.e. unFTP in our case? If not then should we even distinguish between bad user and bad password?

Reply:
I agree, but that is not what we do currently. We wait 1,5 seconds for a bad password but not for a bad ip. This is a security problem, because it reveals the internal reason why the password was not accepted.

Reply:
So I guess we could remove that 1,5 second penalty here and move it into unFTP.

Reply:
Fixed. See update.